src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -241,13 +242,62 @@ ColumnID AbstractLQPNode::get_column_id(const AbstractExpression& expression) co
   return *column_id;
 }
 
+ExpressionUnorderedSet AbstractLQPNode::find_column_expressions(const std::unordered_set<ColumnID>& column_ids) const {
+  const auto& output_expressions = this->output_expressions();
+  auto column_expressions = ExpressionUnorderedSet{};
+
+  for (const auto& output_expression : output_expressions) {
+    const auto column_expression = dynamic_pointer_cast<LQPColumnExpression>(output_expression);
+    if (column_expression && column_ids.contains(column_expression->original_column_id) &&
+        *column_expression->original_node.lock() == *this) {
+      const auto [_, success] = column_expressions.emplace(column_expression);
+      Assert(success, "Did not expect multiple column expressions for the same column id.");
+    }
+  }
+
+  return column_expressions;
+}
+
+bool AbstractLQPNode::has_output_expressions(const ExpressionUnorderedSet& expressions) const {
+  const auto& output_expressions_vec = output_expressions();
+  ExpressionUnorderedSet output_expressions{output_expressions_vec.cbegin(), output_expressions_vec.cend()};
+
+  for (const auto& expression : expressions) {
+    if (!output_expressions.contains(expression)) return false;
+  }
+
+  return true;
+}
+
 bool AbstractLQPNode::is_column_nullable(const ColumnID column_id) const {
   // Default behaviour: Forward from input
   Assert(left_input() && !right_input(),
          "Can forward nullability from input iff there is a left input and no right input");
   return left_input()->is_column_nullable(column_id);
 }
 
+bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSet& expressions) const {
+  DebugAssert(!expressions.empty(), "Invalid input. Set of expressions should not be empty.");
+  DebugAssert(has_output_expressions(expressions),
+              "The given expressions are not a subset of the LQP's output expressions.");
+
+  const auto& unique_constraints = this->unique_constraints();
+  if (unique_constraints->empty()) return false;
+
+  // Look for a unique constraint that is solely based on the given expressions
+  for (const auto& unique_constraint : *unique_constraints) {
+    if (unique_constraint.expressions.size() == expressions.size() &&
```

Comment:
This appears not to do what the documentation is saying (or we are talking about different things). Have you tested this with a superset constraint?

Reply:
I think I get your point:

* Table: a, b, c
* Unique constraint for a

`has_unique_constraint({a, b})` should return true, correct?

Currently, the function is used for single expressions only:
```cpp
bool left_operand_is_unique = left_input()->has_matching_unique_constraint({join_predicate->left_operand()});
bool right_operand_is_unique = right_input()->has_matching_unique_constraint({join_predicate->right_operand()});
```

But, changing the logic in favor of supersets should not break anything.

Reply:
Yepp. That's why I made the distinction between "matching" and "exact" constraints. I expect that the former are more useful for optimization rules etc.

Reply:
Changed to 

```cpp
// Look for a unique constraint that is based on a subset of the given expressions
  for (const auto& unique_constraint : *unique_constraints) {
    if (unique_constraint.expressions.size() <= expressions.size() &&
        std::all_of(unique_constraint.expressions.cbegin(), unique_constraint.expressions.cend(),
                    [&expressions](const auto unique_constraint_expression) {
            return expressions.contains(unique_constraint_expression);
        })) {
      // Found a matching unique constraint
      return true;
    }
  }
```

Reply:
Since AggregateNode profits from the above logic as well, I moved it to lqp_utils.hpp/cpp:
```cpp
bool contains_matching_unique_constraint(const std::shared_ptr<LQPUniqueConstraints>& unique_constraints,
                                         const ExpressionUnorderedSet& expressions);
```

AbstractLQPNode looks as follows now:

```cpp
bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSet& expressions) const {
  DebugAssert(!expressions.empty(), "Invalid input. Set of expressions should not be empty.");
  DebugAssert(has_output_expressions(expressions),
              "The given expressions are not a subset of the LQP's output expressions.");

  const auto& unique_constraints = this->unique_constraints();
  if (unique_constraints->empty()) return false;

  return contains_matching_unique_constraint(unique_constraints, expressions);
}
```